### PR TITLE
convertDoxygen: Add docstrings to more python objects

### PR DESCRIPTION
**Pull request by chadrik**
### Description of Change(s)

With these changes, the `convertDoxgen` build script adds docstrings for many additional python objects.  

For example, from Sdf these objects are now documented:

```
SpecType
Specifier
Permission
Variability
AuthoringError
GetNameForUnit
GetUnitFromName
ValueHasValidType
GetTypeForValueTypeName
GetValueTypeNameForValue
ConvertToValidMetadataDictionary
JustCreatePrimAttributeInLayer
CopySpec
ComputeAssetPathRelativeToLayer
ListOpType
CreatePrimInLayer
JustCreatePrimInLayer
CreateVariantInLayer
```

The primary change is that we no longer filter "file" xml type: these are necessary to find info on non-class functions.

In addition, this PR also provides a change necessary to support creation of [python type stubs](https://github.com/PixarAnimationStudios/OpenUSD/issues/196):  the parser now captures function default arguments. 

### Performance impact

Casting a wider net during parsing increased parse time from 25s to 1m45s.  The addition of slots shaved off ~10s and reduced memory usage. 

I've run this through a profiler, and the majority of the time is spent in the xml parsing, with about half of the parsing-time spent in pxr code and the other half in the parser base class.  I think there's room to make this more efficient if we can avoid instantiating `XMLNode` unnecessarily.  I noticed that for every class/function there is a useless "doxygen" root node, so I tried collapsing that down to a single root, but it only saved 10s and it creates some other problems.  

I also tried using `multiprocessing` to speed up parsing, but even distributing the parsing of files across 16 processes did not speed it up -- I suspect this was due to the amount of `DocElement` instances that needed to be serialized and marshaled across to the primary process after parsing, due to the way that the `multiprocessing` module works in python.   If that inference is correct, this approach could be improved by filtering the resulting `DocElement` classes while parsing, as there are many extraneous functions being captured. 

tl;dr  with this change, the doc generation process takes longer, but there's not a simple solution.

### CODE REVIEW

File: docs/python/doxygenlib/cdWriterDocstring.py [hunk 1]
```diff
@@ -52,29 +52,6 @@ class Writer:
     # we can combine property docstrings for getters/setters.
     propertyTable = {}
 
-    # Default constructor that assumes we're processing a single module
-    def __init__(self):
-        #
```
Comment (chadrik) [inline docs/python/doxygenlib/cdWriterDocstring.py:57]: python doesn't support overloads like C++:  you can't simply define a function twice.  This `__init__` was being replaced by the one below it, so calling `Writer()` without args never actually worked.  I've fixed the use of this class in `convertDoxygen`

File: docs/python/doxygenlib/cdWriterDocstring.py [hunk 2]
```diff
@@ -295,11 +272,14 @@ def generate(self, output_file, docElements):
...
+        if not os.path.exists(outputDir):
+            os.makedirs(outputDir)
+
         # output the lines to disk...
         try:
-            logfile = open(output_file, 'w')
-            logfile.write('\n'.join(lines))
-            logfile.close()
+            with open(output_file, 'w') as logfile:
+                logfile.write('\n'.join(lines))
```
Comment (chadrik) [inline docs/python/doxygenlib/cdWriterDocstring.py:282]: use context manager to ensure that the file is closed in case of an exception (also, more idiomatic)

File: docs/python/doxygenlib/cdParser.py [hunk 3]
```diff
@@ -42,13 +42,18 @@ class XMLNode:
     """
     Rrepresent a single node in the XML tree.
     """
+    __slots__ = ("parent", "name", "attrs", "text", "childNodes")
+
```
Comment (chadrik) [inline docs/python/doxygenlib/cdParser.py:46]: Added slots for performance:  we make a _lot_ of these classes.

File: docs/python/doxygenlib/cdDocElement.py [hunk 4]
```diff
@@ -86,6 +97,10 @@ def isRoot(self):
         """Is this doc element the root of the doxygen XML tree?"""
         return self.kind == 'root'
 
+    def isStatic(self):
+        """Is this doc element static?"""
+        return self.static is not None and self.static == 'yes'
+
```
Comment (chadrik) [inline docs/python/doxygenlib/cdDocElement.py:103]: Added `isStatic` and `__repr__` methods for convenience

File: docs/python/convertDoxygen.py [hunk 5]
```diff
@@ -106,11 +106,13 @@
...
     # load and generate the doc strings for the specific module
     for moduleName in moduleList:
+        if not moduleName:
+            continue
         writer = Writer(packageName, moduleName)
         # Parser.traverse builds the docElement tree for all the 
         # doxygen XML files, so we only need to call it once if we're
         # processing multiple modules
-        if (docList == None):
+        if (docList is None):
```
Comment (chadrik) [inline docs/python/convertDoxygen.py:115]: using `is` instead of `==` is considered idiomatic because `None` is a singleton.

**Comment by jesschimein** [comment]
Filed as internal issue #USD-8482

File: .gitignore [hunk 6]
```diff
@@ -1,4 +1,4 @@
 .p4*
 .DS_Store
 .AppleDouble
-
+*.pyc
```
Comment (pmolodo) [inline .gitignore]: Since you're adding this - don't suppose you could add `__pycache__` to the list as well?
Had that queued up in another PR, but if you just add that to your MR, would avoid a merge conflict...

File: docs/python/doxygenlib/cdParser.py [hunk 7]
```diff
@@ -221,8 +226,7 @@ def parseDoxygenIndexFile(self, doxygen_index_file):
                     continue
                 if (kind == "dir"):
                     continue
-                if (kind == "file"):
-                    continue
```
Comment (pmolodo) [inline docs/python/doxygenlib/cdParser.py:225]: So I think this change is the heart of this PR - unfortunately, when we tested it, it had some negative side effects.
Specifically, many of our classes lost their docstrings after running `convertDoxygen.py` with this change.

One example was `pxr.Usd.ZipFileWriter` - in `Usd/__DOC.py`, we now get:

```python
   result["ZipFileWriter"].__doc__ = """"""
```

instead of:

```python
   result["ZipFileWriter"].__doc__ = """
Class for writing a zip file. This class is primarily intended to
support the .usdz file format. It is not a general-purpose zip writer,
as it does not implement the full zip file specification. However, all
files written by this class should be valid zip files and readable by
external zip modules and utilities.

"""
```

We're using Doxygen 1.8.18... which is old, but also newer than the 1.8.14 listed in [VERSIONS.md](https://github.com/PixarAnimationStudios/OpenUSD/blob/release/VERSIONS.md).

@chadrik, can you confirm whether or not you're seeing the same, and what version of Doxygen you're using?

Comment (spiffmon) [reply docs/python/doxygenlib/cdParser.py:225]: Hey @pmolodo , you made us realize we neglected to update VERSIONS.md for the addition of doxygen-awesome, which requires 1.9.x (we are using 1.9.6 internally).  We will be pushing out updates to VERSIONS.md shortly, but can you retry your experiments with a 1.9.x version?

**Comment by chadrik** [comment]
I did test the results before and after my change to ensure proper results
but it’s quite possible that I missed something. I was doing a synthetic
test of just the docstring build rather than a full cmake build so that
could also have led to differences. I will double check this on my end.

File: docs/python/doxygenlib/cdParser.py (hunk 7 above)
Comment (dsyu-pixar) [reply docs/python/doxygenlib/cdParser.py:225]: I tried a test with only removing these lines and comparing build output (using doxygen 1.8.x). The new __DOC.py appears to have duplicate (empty) values for some things, like ZipFileWriter as @pmolodo notes, which end up overwriting the earlier defined values.

**Comment by chadrik** [comment]
I pushed a fix for this.

File: .gitignore (hunk 6 above)
Comment (pixar-oss) [reply .gitignore:4]: Hi @chadrik, is the .gitignore change critical for the main goal of this PR? For tracking changes for USD releases, it helps if all the changes for a PR are related. 

If the .gitignore change isn't specifically needed for this PR, would it be possible to move the .gitignore change (maybe combined with @pmolodo's change) to a separate PR?

Comment (dsyu-pixar) [reply .gitignore]: Hi @chadrik, is the .gitignore change critical for the main goal of this PR? For tracking changes for USD releases, it helps if all the changes for a PR are related.

If the .gitignore change isn't specifically needed for this PR, would it be possible to move the .gitignore change (maybe combined with @pmolodo's change) to a separate PR?

Comment (chadrik) [reply .gitignore]: Removed!

File: docs/python/doxygenlib/cdParser.py (hunk 7 above)
Comment (chadrik) [reply docs/python/doxygenlib/cdParser.py:225]: This is fixed.
